nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala

```diff
@@ -1540,12 +1540,30 @@ abstract class NirCodeGen
           val offset = genExpr(offsetp, ptr)
           offset withOp Op.Elem(ty, ptr.value, Seq(offset.value))
 
-        case (PTR_SUB, Seq(offsetp, tagp)) =>
-          val st     = unwrapTag(tagp)
-          val ty     = genType(st, box = false)
-          val offset = genExpr(offsetp, ptr)
-          val neg    = negateInt(offset.value, offset)
-          neg withOp Op.Elem(ty, ptr.value, Seq(neg.value))
+        case (PTR_SUB, Seq(argp, tagp)) =>
+          argp match {
```

Comment:
You need to dispatch on `argp.tpe.typeSymbol` here. That would give you a symbol to the type of the expression. For example:

```
val sym = argp.tpe.typeSymbol
sym match {
  case IntClass => /* ... */
  case PtrClass => /* ... */
}
```